Add trainer held item randomization for Gen 3+

spaceonaut opened the pull request:
The goal of this PR is to make sure I'm on the right track with this implementation of #161 (e.g. I didn't forget to change something).

Some rationale for the changes - I named the Gen4 constants (using regexes to convert the linked table of data) instead of just using a range of numbers because 1) there weren't a few small ranges I could have easily used to get "useful" held items and 2) The plan is to add a smarter subset of items, in which case the named constants should be easier to read.

There are a few small changes/comments I added along the way as I was figuring things out. I hope they help others who come along.

I tested this locally on Soul Silver.

One thing I did note was a crash or two as follows that I'm not sure if I'm causing or exacerbating somehow (running on Windows 10):
```
com.dabomstew.pkrandom.exceptions.RandomizerIOException: java.io.FileNotFoundException: C:\Users\Me\Documents\GitHub\universal-pokemon-randomizer-zx\out\production\tmp_soulsilver_orig\a055 (The requested operation cannot be performed on a file with a user-mapped section open)
	at com.dabomstew.pkrandom.romhandlers.Gen4RomHandler.setTrainers(Gen4RomHandler.java:2256)
	at com.dabomstew.pkrandom.romhandlers.AbstractRomHandler.randomizeTrainerHeldItems(AbstractRomHandler.java:1943)
	at com.dabomstew.pkrandom.Randomizer.randomize(Randomizer.java:340)
	at com.dabomstew.pkrandom.newgui.NewRandomizerGUI.lambda$performRandomization$105(NewRandomizerGUI.java:883)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.io.FileNotFoundException: C:\Users\Me\Documents\GitHub\universal-pokemon-randomizer-zx\out\production\tmp_soulsilver_orig\a055 (The requested operation cannot be performed on a file with a user-mapped section open)
	at java.io.FileOutputStream.open0(Native Method)
	at java.io.FileOutputStream.open(FileOutputStream.java:270)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:213)
	at java.io.FileOutputStream.<init>(FileOutputStream.java:162)
	at com.dabomstew.pkrandom.newnds.NDSFile.writeOverride(NDSFile.java:94)
	at com.dabomstew.pkrandom.newnds.NDSRom.writeFile(NDSRom.java:563)
	at com.dabomstew.pkrandom.romhandlers.AbstractDSRomHandler.writeFile(AbstractDSRomHandler.java:214)
	at com.dabomstew.pkrandom.romhandlers.AbstractDSRomHandler.writeFile(AbstractDSRomHandler.java:205)
	at com.dabomstew.pkrandom.romhandlers.AbstractDSRomHandler.writeNARC(AbstractDSRomHandler.java:153)
	at com.dabomstew.pkrandom.romhandlers.Gen4RomHandler.setTrainers(Gen4RomHandler.java:2238)
	... 4 more
```

Ajarmar commented on src/com/dabomstew/pkrandom/romhandlers/Gen4RomHandler.java:
```diff
@@ -4243,4 +4243,42 @@ public BufferedImage getMascotImage() {
     }
+
+    @Override
+    protected void randomizeHeldItems(Trainer t, Settings settings) {
```
The randomization logic for this shouldn't be in `GenXRomHandler` - it should be handled in `AbstractRomHandler`. The method in `AbstractRomHandler` should extract the data using the generation specific `getTrainers`, then perform the generation agnostic randomization logic, then set the new data using the generation specific `setTrainers`. 

To access things like `GenXConstants.consumableHeldItems` in `AbstractRomHandler`, create a method in `RomHandler` that extracts this in a generation specific way (i.e. `getConsumableHeldItems`, and then you implement it in `Gen4RomHandler` as `return Gen4Constants.consumableHeldItems`, etc.)

This whole method can just be incorporated into `randomizeTrainerHeldItems` in `AbstractRomHandler`, it's not really long enough to warrant a sub-method.

spaceonaut replied:
Done.

Ajarmar commented on src/com/dabomstew/pkrandom/romhandlers/Gen4RomHandler.java:
```diff
@@ -4243,4 +4243,42 @@ public BufferedImage getMascotImage() {
+
+    private boolean trainersPokemonHaveItems(Trainer tr) {
+        return (tr.poketype & 2) == 2;
+    }
```
This seems to work the exact same way in every generation, so this would work better as a method in the `Trainer` class. (Same with the "custom moves" one below.)

Ajarmar commented on src/com/dabomstew/pkrandom/newgui/Bundle.properties:
```diff
@@ -583,4 +583,19 @@ GUI.tmFollowEvolutionsCheckBox.toolTipText=<html>When this is selected and TM co
+GUI.heldItemsAllPokemon.text=All Pokemon
+GUI.heldItemsAllPokemon.tooltip=If this option is selected, a held item will be given to all of an applicable trainer's Pokemon.
+GUI.heldItemsHighestLevel.text=Highest Level Only
+GUI.heldItemsHighestLevel.tooltip=If this option is selected, a held item will be given to a trainer's highest level Pokemon (or one of them, if there are multiple at that level).
```
Please make the variable names in the bundle consistent with how it works for all other GUI elements - `GUI.nameOfElement.text` and `GUI.nameOfElement.toolTipText`. (As in, the actual name of the element's variable in `NewRandomizerGUI`.)

spaceonaut replied:
Done.

Ajarmar reviewed (changes requested):
This is a great start, although we won't merge it in until it's done for all generations so that we don't sit here with an unfinished setting for an indefinite amount of time.

I have some concerns with how the GUI looks with the new additions, primarily the vertical size increase. These are my suggestions for how to fix it:

- I don't think any of the settings should be radio buttons. Just make the default "pure random, all Pokemon" and then have "Consumable Only", "Sensible Items" and "Highest Level Only" as check boxes. They should all fit on one row, so we get rid of a row by doing this. (Additionally, this makes it so that "Consumable Only" and "Sensible Items" can be selected at the same time, which they should be able to.)
- Remove the separators, they don't contribute much.
- Move everything up one row (including the "Double Battle Mode" check box and all the "Additional Pokemon" stuff).

As for your error - I couldn't recreate it, so I'm not sure what could have caused it. The error message seems to indicate that you are overwriting a file that is already open elsewhere in the program. Could you supply a settings string for which the error happens for you?

My code specific comments are below. Feel free to ask if you have questions about anything.

Ajarmar commented:
Oh, and one more thing I forgot to mention: [I pushed a change to the GUI a couple of hours ago](https://github.com/Ajarmar/universal-pokemon-randomizer-zx/commit/570bc0d59e3d9bb9dd3ef181f3e626296fa7c9de) that makes it look less weird when a game is loaded, you probably want to merge that into your branch.

spaceonaut commented on src/com/dabomstew/pkrandom/Randomizer.java:
```diff
@@ -336,6 +336,11 @@ public int randomize(final String filename, final PrintStream log, long seed) {
                 break;
         }
 
+        if (settings.getTrainersMod() != Settings.TrainersMod.UNCHANGED) {
```
Done.

spaceonaut commented on src/com/dabomstew/pkrandom/romhandlers/AbstractRomHandler.java:
```diff
@@ -1925,6 +1925,28 @@ public void typeThemeTrainerPokes(Settings settings) {
     }
 
+    @Override
+    public void randomizeTrainerHeldItems(Settings settings) {
```
Done.

spaceonaut commented on src/com/dabomstew/pkrandom/pokemon/TrainerPokemon.java:
```diff
@@ -54,7 +54,11 @@
+        if (heldItem != 0) {
+            s += " [holding 0x"+Integer.toHexString(heldItem)+"]";
+        }
+        return s;
```
Sounds good. I'll probably also replace the RomHandler.randomHeldItem() used for Totem Pokemon with the newly added getAllHeldItems() at some point.

spaceonaut reviewed (commented):
Thank you for the excellent review. I believe I have addressed all requests.

I removed the separators for now, but I would recommend considering them in the event more general trainer checkboxes need to be added. This way, we could make use of this space
![image](https://user-images.githubusercontent.com/82479892/115100835-643c5980-9f0d-11eb-9981-db9a0cfd7682.png)
without confusing users about which settings tie into the additional pokemon/held items.

spaceonaut commented:
Oh, regarding the error I was seeing. I looked up some other errors and found a suggestion to disable anti virus scanning of my output directory. I did so and haven't seen the error crop up again after about a dozen local tests.

spaceonaut commented:
If this still looks good to you, I can try expanding this to other gens (on this same PR, as you requested).

Ajarmar commented on src/com/dabomstew/pkrandom/romhandlers/AbstractRomHandler.java:
```diff
@@ -1925,6 +1924,76 @@ public void typeThemeTrainerPokes(Settings settings) {
+                        highestLevelPoke = tp;
+                        maxLevel = tp.level;
+                    }
+                    tp.heldItem = 0; // set every poke to no item (just in case there was one originally).
```
I think it would be preferable to leave already existing items in place, rather than removing them (after all, the setting is "Add Held Items", not "Set Held Items").

tom-overton replied:
I agree. It also significantly reduces the complexity of handling this for generations 6 and 7, where the randomizer already does some special item handling to ensure Pokemon have the "proper" Mega Stone or Z-Crystal. If you kept the code like this, you would have to check whether one of those "special" items was being held before removing it.

tom-overton replied:
After talking with Ajarmar, I think my concern about Mega Stones and Z-Crystals can simply be solved by adding `hasMegaStone` and `hasZCrystal` properties to the TrainerPokemon class. The `getTrainers` method of the respective RomHandlers would set it based on the vanilla held item. Then, you can ignore held item randomization for Pokemon who have either of these properties set.

spaceonaut replied:
I've changed it so we don't store a new item if there was already an item. I'm not quite sure I follow the Mega Stones and Z-Crystals discussion; in addition to not overwriting already-held items, do I need to check the ZCrystal/mega stone logic? Or does simply not writing a new item if there is one already cover those cases?

tom-overton replied:
Sorry, I think I ended up confusing you with mixed messages. This is what I think should happen, though Ajarmar can give his own ideas as to what should happen:
1. If a Pokemon has a Z-Crystal, don't change their item
2. If a Pokemon has a Mega Stone *and* "Swap Mega Evolvables" for trainer Pokemon is **true**, don't change their item
3. If "Highest Level Only" held item randomization is **true**, and Pokemon who are *not* the highest level have held items, don't change their item and keep it like it was
4. Otherwise, randomize their item

Does that make sense?

Ajarmar reviewed (changes requested):
Thanks for the changes! It looks much better now. I'll see if darkeye wants to give some feedback as well, but I'm satisfied with how this looks now (except for one thing that I missed last time, see below).

Also - the issue with the "fast HP and EXP bars" check box changing names was absolutely not your fault, there was something wrong with the GUI form that made it change name every time a change was made to `Bundle.properties`. I pushed a change that fixes this to master, so you should merge it into your branch so you don't have to deal with that happening again.

tom-overton commented on src/com/dabomstew/pkrandom/newgui/Bundle.properties:
```diff
@@ -583,4 +583,18 @@ GUI.tmFollowEvolutionsCheckBox.toolTipText=<html>When this is selected and TM co
+GUI.tpConsumableItemsOnlyCheckBox.text=Consumable Only
+GUI.tpConsumableItemsOnlyCheckBox.tooltip=<html>Items will be chosen randomly from all single-use held items that a Pokemon can eat/use on its own.<br />This is mostly berries, with a few other items like Focus Sash and White Herb.</html>
+GUI.tpSensibleItemsCheckBox.text=Sensible Items
+GUI.tpSensibleItemsCheckBox.tooltip=<html><b>Not implemented yet.</b><br />Items will be chosen randomly from a subset of the held items that "make sense" for a given Pokemon. For example, Silk Scarf won't be an option for a given Pokemon unless it knows at least one normal move. Another example is that a Yache berry will only be an option on Pokemon that are weak or double weak to ice.</html>
```
I suggest implementing line breaks for this and the Highest Level Only tooltip, since they look much longer than any other tooltip. There's no hard or fast rule for how long a line should be so just experiment; we tend to put them between sentences.

spaceonaut replied:
Done.

tom-overton commented on src/com/dabomstew/pkrandom/newgui/NewRandomizerGUI.java:
```diff
@@ -2584,6 +2626,21 @@ private void romLoaded() {
             tpRegularTrainersCheckBox.setEnabled(false);
             tpRegularTrainersSpinner.setVisible(additionalPokemonAvailable);
 
+            boolean heldItemsSupport = pokemonGeneration == 4; // More gens coming after first review.
```
I would prefer if this was called something like `trainerHeldItemsSupport` because we have multiple different types of held item randomization besides trainers.

spaceonaut replied:
Done.

tom-overton commented on src/com/dabomstew/pkrandom/romhandlers/AbstractRomHandler.java:
```diff
@@ -1925,6 +1924,76 @@ public void typeThemeTrainerPokes(Settings settings) {
+        tp.heldItem = toChooseFrom.get(random.nextInt(toChooseFrom.size()));
+    }
+
+    @Override
```
There's a section *way* at the bottom of AbstractRomHandler.java for "Default Implementations" as the comment calls it. Can you move these three methods here?

spaceonaut replied:
Done.

tom-overton reviewed (commented):
Looks mostly good to me, just some minor things but nothing fundamental. I'd be fine with you starting on Gens 5-7 from here.

spaceonaut commented:
Please take a look. I've added support for Gens 3+ and given a basic test of one game from each gen locally. I rebased all the commits into one so it lands easier (in case of needing to revert).

tom-overton reviewed (commented):
Looks great! Besides some minor things, I just have two questions:
- Is adding support for Gen 2 impossible? I was actually surprised that Gen 3 was possible in the first place, so if it isn't, then no worries; you've already exceeded my expectations as to what's possible.
- Are you implementing the "Sensible Items" settings in this MR?

spaceonaut commented on src/com/dabomstew/pkrandom/constants/Gen3Constants.java:
```diff
@@ -200,6 +200,44 @@
     
     public static final int luckyEggIndex = 0xC5;
 
+    //https://bulbapedia.bulbagarden.net/wiki/List_of_items_by_index_number_(Generation_III)
```
Done

spaceonaut commented on src/com/dabomstew/pkrandom/romhandlers/AbstractRomHandler.java:
```diff
@@ -1925,6 +1924,66 @@ public void typeThemeTrainerPokes(Settings settings) {
+        if (tp.hasMegaStone && settings.isSwapTrainerMegaEvos()) {
+            return; // Don't overwrite mega stones if another setting handled that.
+        }
+        boolean sensibleItemsOnly = settings.isSensibleItemsOnlyForTrainers();
```
Done.

spaceonaut commented:
> Are you implementing the "Sensible Items" settings in this MR?

I was planning on trying that out in a different PR. That way it will be easier to review.

> Is adding support for Gen 2 impossible?

From what I understand, the Gen 2 TrainerData is of fixed size. There are only a handful of trainers with held items in the first place (all Pokefans, it seems). It would not be difficult to randomize the items for these trainers (but probably not interesting). It would probably not be impossible to use the fixed size Trainer Data to remove some held items from the original trainers and put them on others (e.g. Boss trainers), but would likely be difficult. Overall, does not seem worth it to me.

Ajarmar reviewed (changes requested):
Nice work, I just have a few more comments.

spaceonaut commented on src/com/dabomstew/pkrandom/constants/Gen7Constants.java:
```diff
@@ -187,6 +187,44 @@ public static int getPokemonCount(int romType) {
+    public static final int adrenalineOrb = 0x34e;
+    // New non-consumable held items with in-battle NPC effect (not specific to one pokemon family or one move)
+    public static final int terrainExtender = 0x36F, protectivePads = 0x370, electricSeed = 0x371,
+        psychicSeed = 0x372, mistySeed = 0x373, grassySeed = 0x374;
```
Done.

spaceonaut commented on src/com/dabomstew/pkrandom/constants/Gen6Constants.java:
```diff
@@ -237,6 +237,34 @@ public static byte moveCategoryToByte(MoveCategory cat) {
+    // Held items we randomize from Gen 6 are a superset of the held items from Gen 5. Thus, we list only the
+    // new ones here.
+    // New consumable held items.
+    public static final int weaknessPolicy = 0x27f, snowball = 0x289, roseliBerry = 0x2ae,
```
Done.

spaceonaut commented on src/com/dabomstew/pkrandom/constants/Gen5Constants.java:
```diff
@@ -339,6 +339,42 @@ public static Integer getAbsolutePokeNumByBaseForme(int baseForme, int formNum)
+    private static List<Integer> setupAllHeldItems() {
+        List<Integer> list = new ArrayList<>();
+        list.addAll(Gen4Constants.allHeldItems);
+        list.addAll(Arrays.asList(airBalloon, redCard, absorbBulb, cellBattery, ejectButton, fireGem, waterGem,
```
I believe that would add some items (e.g. Gen4 consumables) to the list twice, since they are in the allHeldItems and the consumable items list, right?

Ajarmar replied:
Ah yeah, you're right. Never mind then

spaceonaut commented:
Merge when ready :)

tom-overton reviewed (approved):
LGTM, if Ajarmar also approves we can merge.

Ajarmar reviewed (approved):
Sure, looks good. Initially I wanted everything (including "Sensible Items") in the same PR, but I suspect that all of the Sensible Item handling is going to get quite complicated so sure, let's split it up. I'll merge this in, thanks!
